[FC-0099] feat: add openedx-authz to library apis user_can_create_library and require_permission_for_library_key

MaferMazu opened the pull request:
<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

This PR:
- Introduces [openedx-authz](https://github.com/openedx/openedx-authz) in two api methods `user_can_create_library`
and `require_permission_for_library_key`.
- Modify the collections and publish enforcement to use the new permissions and also check the old one.
- Add tests based on `test_library_permission` but using the new authz system.

Implementation:
We implement `user_has_permission_across_lib_authz_systems` to handle the verification of permissions in both system (the one in `content_libraries/permissions.py` and the new one implemented in `openedx-authz`).

Consequences:
By adding  `openedx-authz` enforcement in  `require_permission_for_library_key`, we are extending the verification in those places: https://github.com/search?q=repo%3Aopenedx%2Fedx-platform%20require_permission_for_library_key&type=code

Important note:
With this change we are adding checks over explicit permissions in the `openedx-authz` system. So we rely on bridgekeeper for the moment to make attribute checking (specific cases: course_creator imply library_creator, and the library having the allow_public_read attribute imply view_library true).

## Supporting information

https://github.com/openedx/openedx-authz/issues/88
https://openedx-authz.readthedocs.io/en/latest/concepts/core_roles_and_permissions/content_library_roles.html#roles-and-permissions-summary-table

## Testing instructions

### Using python test
- I didn't modify the current test, so the tested behavior won't change with this addition.
- I added test that verify the rest api calls when he data uses the new `openedx-authz` system.

### In stage
Stage information: https://openedx.atlassian.net/wiki/spaces/OEPM/pages/5279023110/AuthZ+Stage+Information

**Preparation of the env:**
- Have an library user, a contributor, an author and an admin.
<img width="929" height="845" alt="image" src="https://github.com/user-attachments/assets/f5e2e2f9-ac30-4910-a218-9f74a7b4cc5e" />

Credentials: https://openedx.atlassian.net/wiki/spaces/OEPM/pages/5279023110/AuthZ+Stage+Information

You can create your own library, or use one created. For this I'll use `lib:OpenedX:LIBTESTDELETE/` as example.


**Test:**
**Library User**
- Libraries
    - GET: Enter in https://studio.singapore.edunext.cloud/api/libraries/v2/lib:OpenedX:LIBTESTDELETE/
        - Expected result: Obtain info.
- Library content
    - Edit content: Enter in https://apps.singapore.edunext.cloud/authoring/library/lib:OpenedX:LIBTESTDELETE/
        - Expected result: view the library but be unable to add new elements.
    - Publish: Inside the library click library info
        - Expected result: No see the publish button.
- Collections
    - Edit Collections: Inside https://apps.singapore.edunext.cloud/authoring/library/lib:OpenedX:LIBTESTDELETE/, click collections.
        - Expected result: be unable to edit collections.
 - Delete: You can use this collection for a quick postman test
[simple-delete-37501.json](https://github.com/user-attachments/files/23465637/simple-delete-37501.json)
    - Expected result: 403

**Library Contributor**
- Libraries
    - GET: Enter in https://studio.singapore.edunext.cloud/api/libraries/v2/lib:OpenedX:LIBTESTDELETE/
        - Expected result: Obtain info.
- Library content
    - Edit content: Enter in https://apps.singapore.edunext.cloud/authoring/library/lib:OpenedX:LIBTESTDELETE/
        - Expected result: view the library and be able to add new elements.
    - Publish: Inside the library click library info
        - Expected result: No see the publish button. - This is something we should modify in the front as part of release testing (for now we can view the button but in backend the contributor can not publish that is the expected behavior)
- Collections
    - Edit Collections: Inside https://apps.singapore.edunext.cloud/authoring/library/lib:OpenedX:LIBTESTDELETE/, click collections.
        - Expected result: be able to edit collections.
 - Delete: You can use this collection for a quick postman test (changing the credentials)
[simple-delete-37501.json](https://github.com/user-attachments/files/23465637/simple-delete-37501.json)
    - Expected result: 403

**Library Author**
The same + publish content.

**Library Admin**
It should have access to get, edit the content and collections and delete (soft delete a library)

## Deadline

Ulmo cut.

## Other information

Related to: https://openedx.atlassian.net/wiki/spaces/OEPM/pages/4724490259/PRD+Roles+Permissions

openedx-webhooks commented:
Thanks for the pull request, @MaferMazu!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/rest_api/blocks.py:
```diff
@@ -60,7 +61,9 @@ def get(self, request, lib_key_str):
-        api.require_permission_for_library_key(key, request.user, permissions.CAN_VIEW_THIS_CONTENT_LIBRARY)
+        is_allowed_with_authz = authz_api.is_user_allowed(request.user.username, 'view_library', str(lib_key_str))
+        if not is_allowed_with_authz:
+            api.require_permission_for_library_key(key, request.user, permissions.CAN_VIEW_THIS_CONTENT_LIBRARY)
```
Why not, instead of modifying all these entrypoints, we only modify the `require_permission_for_library_key` method - after some tweaks, so it handles both authorization paths?

MaferMazu replied:
I liked the idea, I applied it. But I needed to maintain a few entrypoints because in the new system we check different things (example: publish, before we only needed "edit", but in the new system we check "publish")

mariajgrimaldi replied:
Does this mean that we're checking twice in some cases?

MaferMazu replied:
Yes, for example, when publishing, we verify the publish using authz. Then, when we check require_permission_for_library_key, we verify, in this case, edit in our system and edit in the old one.

mariajgrimaldi replied:
I wonder if instead of calling these twice we could simply pass a list of permissions, @BryanttV, what do you think? This way we wouldn't have to check twice, but in bulk. Does this make sense to you, @MaferMazu?

MaferMazu replied:
I didn't understand the suggestion of having a list of permissions.

Right now, I am requesting publishing permission (which wasn't accounted for in the old system), and if not, I am using `require_permission_for_library_key` to check whether I have the old permission or the new one with 'edit'. I know we are calling the new permission system twice (once for checking publish and then inside `require_permission_for_library_key`), but I didn't understand how we can avoid that.

mariajgrimaldi replied:
I was wondering if there was a way of asking for a list N of permissions instead of calling the authorization enforcement N times.

MaferMazu replied:
It occurred to me that we could use a different API function to encapsulate those tests (which require a new permission in the new system and a different one in the old), to avoid unnecessary checks. - We can probably improve that, but it is a good starting point.

MaferMazu replied:
I added those checks in `require_authz_lib_permission`.

mariajgrimaldi replied:
There is a batch enforce: https://github.com/casbin/pycasbin/blob/497cb17eae45e24c4f1f24d74175dff64fe7bdc9/casbin/synced_enforcer.py#L161-L166 we could try to implement as one API function to check for publish and edit without calling the enforcer twice. Would that work for this case, @MaferMazu? I don't know if this is really the same as what you're doing.

MaferMazu replied:
I refactored the solution, so we are not calling our enforcement twice.

MaferMazu committed:
feat: add the authz check to the library api function

feat: add the authz publish check in rest_api blocks and containers

feat: add the authz checks in libraries and refactor

feat: add collections checks

feat: update enforcement in serializer file

refactor: refactor the permission check functions

fix: fix value error

fix: calling the queries twice

feanil commented on requirements/edx/github.in:
```diff
@@ -86,3 +86,5 @@
 
 # ... add dependencies here
+
+git+https://github.com/eduNEXT/openedx-authz.git@mfmz/default-policy#egg=openedx-authz==0.5.0
```
Just confirming that this is temporary until the openedx-authz branch lands on the default branch?

MaferMazu replied:
Yes

mariajgrimaldi reviewed (commented):
I know this is still a work in progress, but I think we should add unit tests for the cases in
[https://github.com/openedx/edx-platform/blob/8fad71724e6b87e4c3f2c95f69b21c7e7507a5da/openedx/core/djangoapps/content_libraries/tests/test_content_libraries.py#L565-L745](https://github.com/openedx/edx-platform/blob/8fad71724e6b87e4c3f2c95f69b21c7e7507a5da/openedx/core/djangoapps/content_libraries/tests/test_content_libraries.py#L565-L745)
to make sure we're still covering the same scenarios we're modifying here, assuming these are the ones affected.

MaferMazu commented:
@BryanttV @mariajgrimaldi, following the recommendation to add our enforcement in `require_permission_for_library_key`, we are already adding the enforcement in some of the places I identified.

This PR currently covers the parts marked in here: https://github.com/orgs/openedx/projects/86/views/1?pane=issue&itemId=132840240&issue=openedx%7Copenedx-authz%7C88 (tested)

Right now, I am working on the collection permissions (I know I have some tests failing, but I'll work on that tomorrow morning) to try to have all the enforcement in this PR, since we're far from having all the enforcement points updated.

I added `require_authz_lib_permission` to handle the permissions that weren't in the old system.

Things this PR needs:

- [x] Fix collection tests
- [x] Update the remaining enforcement points
- [x] Create a test_content_library_permission using data from the new system

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/api/libraries.py:
```diff
@@ -712,3 +722,42 @@ def get_backup_task_status(
+    return mapping.get(permission, permission)
+
+
+def require_authz_lib_permission(
```
This function is a bit unclear, so two suggestions:
1. Could we rename it? The current name doesn’t give enough context about what it does.
2. The docstring could also use some improvement to better explain its purpose.

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/api/libraries.py:
```diff
@@ -712,3 +722,42 @@ def get_backup_task_status(
+        or authz_api.is_user_allowed(
+            user,
+            permission,
+            str(library_obj),
```
Maybe str(library_key) would be better than str(library_obj)

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/rest_api/collections.py:
```diff
@@ -198,6 +215,12 @@ def update_items(self, request: RestRequest, *args, **kwargs) -> Response:
         Collection and items must all be part of the given library/learning package.
         """
         content_library = self.get_content_library()
+        key = LibraryLocatorV2.from_string(str(content_library))
```
If using the content_library object is far more common than the key, maybe we could move this to the function below instead

mariajgrimaldi reviewed (changes requested):
Here's another round of review. My main concern is transparency and readability. It's a bit confusing to have multiple functions checking permissions, so we might need to rethink how to centralize all checks in one place. That way, we avoid maintaining different versions and make it easier for others to follow the authorization logic.

I want to prioritize parity functionality instead of quality and nits, so we can find a good place in between.

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/api/libraries.py:
```diff
@@ -232,7 +235,14 @@ def user_can_create_library(user: AbstractUser) -> bool:
+    has_perms = user.has_perm(library_permission) or authz_api.is_user_allowed(
+        user,
+        lib_permission_in_authz,
+        'sc:generic',
```
What does `sc:generic` mean in this context?

MaferMazu replied:
It means global permission, but I'll use '*'.
Reference: https://github.com/openedx/openedx-authz/issues/131

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/api/libraries.py:
```diff
@@ -712,3 +722,81 @@ def get_backup_task_status(
+        permissions.CAN_VIEW_THIS_CONTENT_LIBRARY: 'view_library',
+        permissions.CAN_VIEW_THIS_CONTENT_LIBRARY_TEAM: 'view_library_team',
+    }
+    return mapping.get(permission, permission)
```
nit: if we're not making this a constant it might be best to return right away instead of using the mapping variable. Same comment below.

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/rest_api/libraries.py:
```diff
@@ -219,7 +219,7 @@ def post(self, request):
         Create a new content library.
         """
-        if not request.user.has_perm(permissions.CAN_CREATE_CONTENT_LIBRARY):
+        if not api.user_can_create_library(request.user):
```
What did we change about this function? I'm not really sure what we changed regarding the new system.

MaferMazu replied:
It only allows to check in our system if someone has the create_library permission. But, for now, it doesn't change anything because we don't have in our system a role with that permission.

It will have an effect when we apply the change of being able to have attribute checking for create_library (verify if they have course_creator)

mariajgrimaldi replied:
So where do we specify how this is currently working? I think this is related to this comment: https://github.com/openedx/edx-platform/pull/37501/files#r2499357506. I think it'd be useful to leave detailed comments on how this is currently working and the known issues we're introducing.

MaferMazu replied:
https://github.com/openedx/edx-platform/pull/37501/commits/f9039a3e52cb155b70ee621fc0dc239b63c63e8c

mariajgrimaldi commented on requirements/edx/base.txt:
```diff
@@ -33,13 +33,16 @@ asgiref==3.10.0
 asn1crypto==1.5.1
     # via snowflake-connector-python
+async-timeout==5.0.1
+    # via redis
```
Can you run `make upgrade-package package=openedx-authz` to use the latest version?

rodmgwgu commented on openedx/core/djangoapps/content_libraries/rest_api/blocks.py:
```diff
@@ -238,7 +238,7 @@ def post(self, request, usage_key_str):
             key.lib_key,
             request.user,
-            permissions.CAN_EDIT_THIS_CONTENT_LIBRARY
+            'publish_library_content'
```
Should we have this in a constant instead?

MaferMazu replied:
I added this PR to be able to use constants: https://github.com/openedx/openedx-authz/pull/125

rodmgwgu commented on openedx/core/djangoapps/content_libraries/rest_api/collections.py:
```diff
@@ -51,16 +51,10 @@ def get_content_library(self) -> ContentLibrary:
 
         lib_key_str = self.kwargs["lib_key_str"]
         library_key = LibraryLocatorV2.from_string(lib_key_str)
-        permission = (
```
Not sure why this was like this before, but by removing this we are changing existing behavior, I think it would be better to keep behavior as-is, unless you are sure this won't affect anything?

MaferMazu replied:
I'll follow your suggestion taking into account that we are managing this with the new permissions in `user_has_permission_across_lib_authz_systems`.

rodmgwgu commented on openedx/core/djangoapps/content_libraries/api/libraries.py:
```diff
@@ -232,7 +235,14 @@ def user_can_create_library(user: AbstractUser) -> bool:
-    return user.has_perm(permissions.CAN_CREATE_CONTENT_LIBRARY)
+    library_permission = permissions.CAN_CREATE_CONTENT_LIBRARY
+    lib_permission_in_authz = _transform_legacy_lib_permission_to_authz_permission(library_permission)
+    has_perms = user.has_perm(library_permission) or authz_api.is_user_allowed(
```
Should we differentiate between legacy permissions and new ones? I'm thinking on an edge case where a permission may have the same name in both new and old permission systems, would this be possible?

bmtcril replied:
How are legacy permissions going to be managed going forward?

MaferMazu replied:
> Should we differentiate between legacy permissions and new ones? I'm thinking on an edge case where a permission may have the same name in both new and old permission systems, would this be possible? - @rodmgwgu 

For now it won't happend because the old permission contains a namespace `'content_libraries.<permission_name>'`, and what we use in the checks for authz is only the `'<permission_name>'`, - So for the MVP and what is defined the permissions, they won't have the same name.

But for future migration in the new system, this is something to think about, and it depends on the rules for naming we decide to manage in the `openedx-authz` permissions.

References:

- https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/content_libraries/permissions.py
- https://github.com/openedx/openedx-authz/blob/main/openedx_authz/constants/permissions.py#L9

MaferMazu replied:
> How are legacy permissions going to be managed going forward? - @bmtcril 

For this release we still check them, and for future release the idea is to remove them, to apply the expand/contract pattern to change the libraries permissions system.

Reference: https://github.com/openedx/edx-platform/issues/37409

mariajgrimaldi replied:
@MaferMazu: I think in openedx-authz we use the namespace as part of the scope, so I'm not sure if we should revisit adding it to the permission. Do you see a specific use case for this? If so, can you open an issue to do a proper follow-up on this?

MaferMazu replied:
@mariajgrimaldi I created a issue related to this https://github.com/openedx/openedx-authz/issues/136, but I think it is something we can hanlde out of the mvp.

bmtcril replied:
Just a note that I think if we make this change after MVP we will need another migration to update existing names. Does that sound right?

mariajgrimaldi replied:
yes

bmtcril commented on openedx/core/djangoapps/content_libraries/rest_api/collections.py:
```diff
@@ -144,6 +143,11 @@ def partial_update(self, request: RestRequest, *args, **kwargs) -> Response:
+        api.require_permission_for_library_key(
+            content_library.library_key,
+            request.user,
+            'edit_library_collection'
```
It would be great if we could make all of these constants

MaferMazu replied:
I created this PR to be able to use constants: https://github.com/openedx/openedx-authz/pull/125

bmtcril reviewed (commented):
Nothing to add to the other reviews aside from a general preference to make strings constants to prevent accidental misspellings etc.

MaferMazu committed:
test: add structure for test and apply feedback

refactor: refactor the tests and apply feedback

fix: apply feedback

Revert "refactor: refactor the tests and apply feedback"

This reverts commit aa0bd527dd7bc7dec4a7ad7adb41a3c932f4a587.

refactor: use constants and avoid mapping

test: fix the test to have them in order

docs: about we rely on bridgekeeper and the old check for two cases

docs: update openedx/core/djangoapps/content_libraries/api/libraries.py

Co-authored-by: Maria Grimaldi (Majo) <maria.grimaldi@edunext.co>

refactor: use global scope wildcard instead of *

refactor: allow receiving PermissionData objects

refactor: do not inherit from BaseRolesTestCase to favor CL setup methods

If both BaseRolesTestCase and ContentLibrariesRestApiTest define a method
with the same name (e.g., setUp()), Python will use the one found first
in the MRO, which is the one in BaseRolesTestCase because it is
listed first in the class definition leading to unexpected behavior.

refactor: remove unnecessary imports and indent

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/api/blocks.py:
```diff
@@ -4,6 +4,7 @@
 These methods don't enforce permissions (only the REST APIs do).
 """
 from __future__ import annotations
+
```
Can we drop these changes? They're only style changes I think.

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/api/libraries.py:
```diff
@@ -750,3 +756,79 @@ def get_backup_task_status(
+        legacy_permission = _transform_authz_permission_to_legacy_lib_permission(permission)
+    return (
+        # Check both the legacy and the new openedx-authz permissions
+        user.has_perm(perm=legacy_permission, obj=library_obj)
```
Remember we should include here a detailed comment about how we're relying on bridgekeeper to check the specific rules we haven't implemented on our end (the attribute checks).

MaferMazu replied:
I added the info in the function doc.

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/tests/test_content_libraries.py:
```diff
@@ -1244,3 +1249,313 @@ def test_xblock_handler_invalid_key(self):
+
+        return user
+
+    @ddt.data(
```
I haven't gone through this tests in details, I can do it later. Could you confirm if we're testing at least [these test cases](https://github.com/openedx/edx-platform/pull/37501#pullrequestreview-3370026720)? Thanks!

MaferMazu replied:
Yes, but I am not checking the team managment apis (because we are handle it in `openedx-authz`)

MaferMazu replied:
Btw, I had to change the test structure (ddt and isolation), and use a structure more similar to the one you mention (sequential), to avoiding several calls to the casbin_table and inconsistencies between tests, because the operations I'm using makes changes in the db.

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/api/libraries.py:
```diff
@@ -245,7 +244,18 @@ def user_can_create_library(user: AbstractUser) -> bool:
+    # Check both legacy and authz permissions
+    # Currently the authz system only check for this specific permission if is superuser or staff
+    # so we rely on the legacy permission check the course_creator status, that is the current condition
+    # to allow a user to create a content library.
```
```suggestion
# The authz_api.is_user_allowed check only validates permissions within a specific library context. Since 
# creating a library is not tied to an existing one, we use user.has_perm (via Bridgekeeper) to check if the user 
# can create libraries, meaning they have the course creator role. In the future, this should rely on a global (*) 
# role defined in the Authorization Framework for instance-level resource creation.
```

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/api/libraries.py:
```diff
@@ -750,3 +760,87 @@ def get_backup_task_status(
+    """
+    # Right now, we don't have a permission defined or a role that grants can_create_content_library
+    # using openedx-authz, so we just return the same permission, and in openedx-authz we are only checking
+    # for superusers/staff for that permission.
```
```suggestion
# There is no dedicated permission or role for can_create_content_library in openedx-authz yet, 
# so we reuse the same permission to rely on user.has_perm via Bridgekeeper.
```

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/rest_api/collections.py:
```diff
@@ -110,6 +110,11 @@ def create(self, request: RestRequest, *args, **kwargs) -> Response:
         Create a Collection that belongs to a Content Library
         """
         content_library = self.get_content_library()
+        api.require_permission_for_library_key(
+            content_library.library_key,
+            request.user,
+            authz_permissions.CREATE_LIBRARY_COLLECTION.identifier
+        )
```
Two comments:

1. [nit] I’m not a fan of using `.identifier` here since it’s not very clear what it represents. Could we use the permission object instead and handle this logic inside `requirement_permission_for_library_key`?
2. Just to confirm, were no permissions being checked here before?

MaferMazu replied:
> 1. [nit] I’m not a fan of using .identifier here since it’s not very clear what it represents. Could we use the permission object instead and handle this logic inside requirement_permission_for_library_key?

I am using the identifier because that api method receives a str (and makes sense for the old checks). If I changed that to obtain an object—for example, an AuthZPermission—I would need to update all the checks that receive `permissions.NAME` and change it (and apply the logic of transformation between systems outside a unified function)

> 2. Just to confirm, were no permissions being checked here before?

Yes, the CRUD over collection are new permissions, and also the publish. 

https://github.com/openedx/edx-platform/blob/7717ec8a321bf699e834c7e978f2d3679b4a7655/openedx/core/djangoapps/content_libraries/api/libraries.py#L781-#L790

mariajgrimaldi replied:
I don't want to block this further, but this could take `require_permission_for_library_key(..., permission: str | PermissionData)` and distinguish between the two: https://github.com/eduNEXT/edx-platform/blob/a35cfc243a3394fb225f156cdc4b35c0633a7aa6/openedx/core/djangoapps/content_libraries/api/libraries.py#L332-L349, no need to change anything else, I think.

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/api/libraries.py:
```diff
@@ -750,3 +760,87 @@ def get_backup_task_status(
+    - CAN_CREATE_CONTENT_LIBRARY, because the condition to create a library is being a course creator,
+      and openedx-authz doesn't have that attribute check implemented yet.
+    - CAN_VIEW_THIS_CONTENT_LIBRARY, regarding the libraries with the allow_public_read attribute,
+      because openedx-authz doesn't have that attribute check implemented yet.
```
```suggestion
Temporary: this function uses Bridgekeeper-based logic for cases not yet modeled in openedx-authz.

Current gaps covered here:
- CAN_CREATE_CONTENT_LIBRARY: we call user.has_perm via Bridgekeeper to verify the user is a course creator.
- CAN_VIEW_THIS_CONTENT_LIBRARY: we respect the allow_public_read flag via Bridgekeeper.

Replace these with authz_api.is_user_allowed once openedx-authz supports these conditions natively (including global (*) roles).
```

mariajgrimaldi reviewed (commented):
LGTM! Thanks for addressing all my comments :) I tested several cases on our sandbox, and everything seems to work as expected.

What’s still missing, though, is a list of test cases we should run on the sandbox, could you please attach them? I'll leave my approval once I've tested them!

I also left a few suggestions for improving inline comments for clarity. Thanks for all the hard work!

MaferMazu commented:
This is ready for re-review @rodmgwgu @bmtcril @mariajgrimaldi.

Note: Before merging this, I would like to test it, with a branch in authz with the namespaces. Only to make sure everything works well. It should, but only to confirm.

mariajgrimaldi commented:
@MaferMazu: friendly ping about my previous comment here https://github.com/openedx/edx-platform/pull/37501#pullrequestreview-3433603666

I see you updated the cover letter so I'll be going through that :)

mariajgrimaldi commented on openedx/core/djangoapps/content_libraries/api/libraries.py:
```diff
@@ -245,7 +244,18 @@ def user_can_create_library(user: AbstractUser) -> bool:
+    has_perms = user.has_perm(library_permission) or authz_api.is_user_allowed(
+        user,
+        lib_permission_in_authz,
+        '*',
```
You can use this: https://github.com/openedx/openedx-authz/blob/main/openedx_authz/api/data.py#L34

mariajgrimaldi commented:
@MaferMazu: can we report this in the openedx-authz repo?

> Expected result: No see the publish button. - This is something we should modify in the front as part of release testing (for now we can view the button but in backend the contributor can not publish that is the expected behavior)

mariajgrimaldi reviewed (approved):
LGTM! 

Thank you so much for the hard work. I left a few nits for you to consider, but everything else looks good to go. Everything I tested also looked good! Thanks :)

MaferMazu commented:
I am having problems with the test using the last `openedx-authz`. I'm working on it.

mariajgrimaldi commented:
@MaferMazu could you check these changes to see if they help with the unit tests? [https://github.com/openedx/edx-platform/commit/3cd4176823ccd6b320488f3cba4caab80aba5410](https://github.com/openedx/edx-platform/commit/3cd4176823ccd6b320488f3cba4caab80aba5410)

A few notes:

1. The atomic blocks aren’t needed since the API we call already handles atomic consistency.
2. There was an issue when inheriting from classes that both define the same methods (like `setUp`, `setUpClass`, etc.) and share the same parent (Django test class), which caused unexpected behavior.
3. I split the test into smaller, clearer blocks to make debugging and review easier.

Let me know what you think of the changes. Thanks!

bmtcril reviewed (approved):
I've tested this locally and it looks good to me once @mariajgrimaldi 's comments are sorted!

MaferMazu commented:
@mariajgrimaldi, thanks a lot for the help with the test. I applied your changes, and they work. Thanks again.

The last change I did was to upgrade the openedx-authz package.

@mariajgrimaldi, can you help me review this?

mariajgrimaldi reviewed (approved):
LGTM! Thank you :)

MaferMazu committed:
chore: bump openedx-authz version
